### DISCUSSION TITLE
AS5600 scratch?

### DISCUSSION THREAD

**Discussion by roguestreak** [marcel-licence/ml_synth_sampler_example#4] [General]
Hi Marcel
After several years of watching I STILL haven't gotten around to making a synth, I recently got myself a 3d printer and thought I might start with the bare basics of the sample scratch with AS5600 (scratch assembly is printed and ready);
- is ml_synth_sampler_example now the best way of doing this, with esp32_midi_sampler being obsoleted?  Or stick with the latter?  I note that this project doesn't have the "#define AS5600_ENABLED" in the config.h, would it work just adding it or is not not supported here?
- Is the rp2040 appropriate for this?  I don't need high fidelity, just PWM is fine for this.  I've got a few generic rp2040 boards lying around with I think 16MB memory, also somewhere have some ESP32 but I think the ones I got had low or no PSRAM?

Maybe after that I might get around to getting synth sounds out of it..
Thanks, Lee

**Comment by marcel-licence** [comment]
Hi Lee,

I planned to bring all features of the esp32_midi_sampler into ml_synth_sampler_example.
The rp2040 should be capable of this job. Instead of pushing the sample into the RAM it could be played back directly from FLASH. I've prepared a feature for FLASH sample playback within the last days. This would also work for the ESP32 without PSRAM. 
I will look at this soon.
If you like I will create a branch on GitHub for testing.

Best regards
Marcel

**Reply by roguestreak** [reply]
Thanks Marcel, no rush, I've got to go back and read over everything I need to setup, get the arduino IDE working again, etc

**Reply by marcel-licence** [reply]
Hi Lee,
please try out the following project: https://github.com/marcel-licence/ml_scratch
It works completely from flash. Some features are missing but it is good enough to test out the AS5600 and the audio output. Samples need to be replaced by own files because they are empty at the moment to avoid any copyright strikes.
Best regards,
Marcel

**Reply by roguestreak** [reply]
Thanks Marcel, I'll give that one a crack

**Reply by roguestreak** [reply]
Hi Marcel, is the above link only for ESP32 as it states or was that leftover text from an old project?  Just checking whether this one does include the RP2040 support or whether I should track down my ESP32 board

**Reply by marcel-licence** [reply]
Hi, I tested it on ESP32. I updated the project some minutes ago to support the RP2040. It is not tested yet.
It is compilable and an update of the sample data is required.

**Reply by marcel-licence** [reply]
I was able to test with the RP2350 with some little changes (the project is updated now). That works well. For some reason the RP2040 does not work at all. No USB port, no blinking, nothing. Seems like it does not want to boot any more using the same firmware.

**Reply by marcel-licence** [reply]
To get it running on RP2040 please delete `libML_SynthTools_RP2350.a`. It seems it causes some confusion in the linker. Hope I can sort it out shortly

**Reply by marcel-licence** [reply]
Another update, the lib is updated again and deleting not necessary any more. I run a successful test on the rp2040 and rp2350 and the ESP32.

**Reply by roguestreak** [reply]
Thanks Marcel, I wasn't even aware of the RP2350 until last week!
I've been trawling through your over projects, videos, and comments to see if I can get things straight, can I please confirm;
- no need to define the rp2040, it's just based on your selected board in the Arduino IDE?
- audio pins are wherever the PWM0 pins are, but in this case only one pin rather than the two stereo pins?
- AS5600 connection is via the I2C0 pins?

I've gotten samples converted into .c / .h, how does the multiple sample array work in the current incarnation?  In one of the other projects I see you were using buttons to go up / down and then lock in a sample, does the current incarnation possibly just add them one after the other so if I keep rotating the disk then the next sample will appear?  Couldn't quite decipher what the code was doing.

While trying to find the PWM pinouts I came across this project; if I'm seeing it right he's managing to get the rp2040 to work directly in USB MIDI input mode, to drive the synth directly from a USB MIDI keyboard?  Is there anything usable in there for you for your projects?  Would be great to have the ability to connect directly to a USB MIDI keyboard without an expensive (relatively) intermediate board
https://github.com/risgk/digital-synth-pra32-u

Thanks again for all your hard work over the years
Lee
